mender-update/deployments/platform/boost_log/deployments.cpp

```diff
@@ -0,0 +1,197 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <mender-update/deployments.hpp>
+
+#include <boost/date_time/posix_time/posix_time.hpp>
+#include <boost/log/attributes.hpp>
+#include <boost/log/common.hpp>
+#include <boost/log/sinks.hpp>
+#include <boost/smart_ptr/shared_ptr.hpp>
+
+#include <filesystem>
+#include <string>
+
+#include <common/error.hpp>
+#include <common/io.hpp>
+#include <common/json.hpp>
+#include <common/log.hpp>
+#include <common/path.hpp>
+
+namespace mender {
+namespace update {
+namespace deployments {
+
+using namespace std;
+
+namespace logging = boost::log;
+namespace expr = boost::log::expressions;
+namespace sinks = boost::log::sinks;
+
+namespace fs = std::filesystem;
+
+namespace error = mender::common::error;
+namespace io = mender::common::io;
+namespace json = mender::common::json;
+namespace mlog = mender::common::log;
+namespace path = mender::common::path;
+
+static void JsonLogFormatter(logging::record_view const &rec, logging::formatting_ostream &strm) {
+	strm << "{";
+
+	auto val = logging::extract<boost::posix_time::ptime>("TimeStamp", rec);
+	if (val) {
+		strm << R"("timestamp":")"
+			 << json::EscapeString(boost::posix_time::to_iso_extended_string(val.get())) << "\",";
+	}
+
+	auto level = logging::extract<mlog::LogLevel>("Severity", rec);
+	if (level) {
+		string lvl = mlog::ToStringLogLevel(level.get());
+		strm << R"("level":")" << json::EscapeString(lvl) << "\",";
+	}
+
+	strm << R"("message":")" << json::EscapeString(*rec[expr::smessage]) << "\"}";
+}
+
+static const size_t kMaxExistingLogs = 5;
+static const uintmax_t kLogsFreeSpaceRequired = 100 * 1024; // 100 KiB
+
+error::Error DeploymentLog::PrepareLogDirectory() {
+	fs::path dir_path(data_store_dir_);
+	bool created = fs::create_directories(dir_path);
+	if (created) {
+		// should basically never happen, but if we happened to create the
+		// directory, it's empty and thus well-prepared
+		return error::NoError;
+	}
+
+	vector<string> old_logs;
+	fs::directory_iterator end_itr; // default construction yields past-the-end
+	for (fs::directory_iterator itr(dir_path); itr != end_itr; ++itr) {
+		fs::path file_path = itr->path();
+		if (!fs::is_regular_file(file_path)) {
+			continue;
+		}
+
+		string file_name = file_path.filename().string();
+
+		if (file_name == "deployments.0000." + id_ + ".log") {
```

Comment:
Q: Lots and lots of string comparisons. Why not pattern/regex match it in some way?

Reply:
I would try to avoid regex if possible. It's typically a big dependency, and I suspect we have to remove it when aiming for smaller platforms. It is used in a few other places already, but let's try to avoid increasing its usage unless the alternative is very messy.

Reply:
Yeah, I remember you had issues with it before so I tried to avoid it.